src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -18,169 +18,59 @@
-                         "Can't use the NonZeroElements class with a scalar shape");
-        }
-
-        template <typename T>
-        void find_indices(const T* values)
-        {
-            m_current_index = Shape(m_input_shape.size(), 0UL);
-            const auto values_count = shape_size(m_input_shape);
-
-            const T zero_value = T{0};
-            for (size_t i = 0; i + 1 < values_count; ++i)
-            {
-                if (values[i] != zero_value)
-                {
-                    add_to_results(m_current_index);
-                }
-
-                next_index();
-            }
-
-            // check the last element in the input values
-            if (values_count != 0 && values[values_count - 1] != zero_value)
-            {
-                add_to_results(m_current_index);
-            }
-        }
+    const auto input_shape = data->get_shape();
+    size_t input_rank = input_shape.size();
+    const T* input_values = data->get_data_ptr<T>();
+    Shape out_shape;
 
-        void generate_all_indices()
-        {
-            m_current_index = Shape(m_input_shape.size(), 0UL);
-            size_t i = 0;
-            const auto values_count = shape_size(m_input_shape);
-            while (i + 1 < values_count)
-            {
-                add_to_results(m_current_index);
-                next_index();
-                ++i;
-            }
-            add_to_results(m_current_index);
-        }
+    size_t non_zero_count = runtime::reference::non_zero_get_count<T>(input_values, input_shape);
+    size_t out_elem_count = (input_rank == 0) ? non_zero_count : (input_rank * non_zero_count);
 
-        const Results_t& get_indices() const { return m_results; }
-    private:
-        /// \brief Adds single dimensions of an index into the matching element of the results
-        void add_to_results(const Shape& index)
-        {
-            for (size_t dim = 0; dim < index.size(); ++dim)
-            {
-                m_results.at(dim).push_back(index[dim]);
-            }
-        }
+    if (out_elem_count == 0)
+    {
+        out_shape = Shape{0};
+    }
+    else if (input_rank == 0)
+    {
+        out_shape = Shape{1, 1};
+    }
+    else
+    {
+        out_shape = Shape{input_rank, non_zero_count};
+    }
 
-        // Generates an index pointing to the next element in the flattened tensor
-        // It behaves similar to flipping bits when incrementing a binary number
-        inline void next_index()
-        {
-            for (size_t dim = m_current_index.size() - 1; dim >= 0; --dim)
-            {
-                auto& dim_value = m_current_index.at(dim);
-                if (dim_value + 1 == m_input_shape[dim])
-                {
-                    dim_value = 0;
-                }
-                else
-                {
-                    ++dim_value;
-                    return;
-                }
-            }
-        }
+    std::vector<U> output_vec(out_elem_count);
```

Comment:
Don't use vector, use AlignedBuffer. Your old code had a bug that was not related to MacOS and would have triggered AddressSanitizer if you had run that. In your code
```
    U* data_ptr = nullptr;
    if (out_elem_count > 0)
    {
        runtime::AlignedBuffer buffer(out_elem_count * sizeof(U));
        data_ptr = buffer.get_ptr<U>();
    }

    // data_ptr is invalid here since it is outsize of the scope where AlignedBuffer was created.
```
AlignedBuffer MUST be outside the curly braces. AlignedBuffer is deleted on exit of the `if` condition so MacOS or Linux data_ptr is invalid.

Reply:
Thanks! updated.